Q: Why is iOS not supported?

`malgo` does not list iOS as a supported platform, but macOS. As far as I understand, Miniaudio is based on Core Audio for both macOS and iOS. So I am surprised by this restriction.
Is this a hard constraint, i.e. is there a technical burden that prohibits deployment, or is it just not tested?

Many thanks for an answer to help me out!

@gen2brain,
Thanks for the response. What should be in the `.m` file? Any draft would be great. I tried to use `malgo` with [fyne](https://github.com/fyne-io/fyne) and cross-compile failed with errors like

```bash
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSZone.h:19:63: error: unknown type name 'NSString'; did you mean 'GoString'?
```

I tried to hard code `miniaudio.go` by adding some more flags:

```go
// Package malgo - Mini audio library (miniaudio cgo bindings).
package malgo

/*
#cgo CFLAGS: -std=gnu99 -Wno-unused-result
#cgo ma_debug CFLAGS: -DMA_DEBUG_OUTPUT=1

#cgo linux,!android LDFLAGS: -ldl -lpthread -lm
#cgo linux,arm LDFLAGS: -latomic
#cgo openbsd LDFLAGS: -ldl -lpthread -lm
#cgo netbsd LDFLAGS: -ldl -lpthread -lm
#cgo freebsd LDFLAGS: -ldl -lpthread -lm
#cgo android LDFLAGS: -lm

#cgo !noasm,!arm,!arm64 CFLAGS: -msse2
#cgo !noasm,arm,arm64 CFLAGS: -mfpu=neon -mfloat-abi=hard
#cgo noasm CFLAGS: -DMA_NO_SSE2 -DMA_NO_AVX2 -DMA_NO_AVX512 -DMA_NO_NEON

#cgo CFLAGS: -x objective-c
#cgo LDFLAGS: -framework Foundation
#cgo LDFLAGS: -framework AudioUnit -framework CoreAudio -framework AudioToolbox
#include "foundation.h"
#import <Foundation/Foundation.h>

#include "malgo.h"
*/
import "C"

// SampleSizeInBytes retrieves the size of a sample in bytes for the given format.
func SampleSizeInBytes(format FormatType) int {
	cformat := (C.ma_format)(format)
	ret := C.ma_get_bytes_per_sample(cformat)
	return int(ret)
}

const (
	rawDeviceInfoSize = C.sizeof_ma_device_info
)

```

but didn't work. There were heaps of errors. Some like:

```bash
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSURLSession.h:1441:1: error: expected identifier or '('
@end
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSURLSession.h:1457:1: error: expected identifier or '('
@property (copy, readonly) NSDateInterval *taskInterval;
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSURLSession.h:1462:1: error: expected identifier or '('
@property (assign, readonly) NSUInteger redirectCount;
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSURLSession.h:1465:1: error: expected external declaration
- (instancetype)init API_DEPRECATED("Not supported", macos(10.12,10.15), ios(10.0,13.0), watchos(3.0,6.0), tvos(10.0,13.0));
^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSURLSession.h:1465:4: error: redefinition of 'id' as different kind of symbol
- (instancetype)init API_DEPRECATED("Not supported", macos(10.12,10.15), ios(10.0,13.0), watchos(3.0,6.0), tvos(10.0,13.0));
   ^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:231:22: note: expanded from macro 'instancetype'
#define instancetype id
                     ^
```

Hope you can point me to the right direction. Thanks.

@gen2brain,

If drop `AudioUnit`, seem to go forward, but got the following linking error:

```bash
/usr/local/go/pkg/tool/darwin_amd64/link: running /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang failed: exit status 1
Undefined symbols for architecture arm64:
  "_AVAudioSessionCategoryAmbient", referenced from:
      _ma_to_AVAudioSessionCategory in 000041.o
  "_AVAudioSessionCategoryMultiRoute", referenced from:
      _ma_to_AVAudioSessionCategory in 000041.o
  "_AVAudioSessionCategoryPlayAndRecord", referenced from:
      _ma_context_init__coreaudio in 000041.o
      _ma_to_AVAudioSessionCategory in 000041.o
  "_AVAudioSessionCategoryPlayback", referenced from:
      _ma_context_init__coreaudio in 000041.o
      _ma_to_AVAudioSessionCategory in 000041.o
  "_AVAudioSessionCategoryRecord", referenced from:
      _ma_context_init__coreaudio in 000041.o
      _ma_to_AVAudioSessionCategory in 000041.o
  "_AVAudioSessionCategorySoloAmbient", referenced from:
      _ma_to_AVAudioSessionCategory in 000041.o
  "_AVAudioSessionInterruptionNotification", referenced from:
      -[ma_ios_notification_handler init:] in 000041.o
      -[ma_ios_notification_handler remove_handler] in 000041.o
  "_AVAudioSessionInterruptionTypeKey", referenced from:
      -[ma_ios_notification_handler handle_interruption:] in 000041.o
  "_AVAudioSessionRouteChangeNotification", referenced from:
      -[ma_ios_notification_handler init:] in 000041.o
      -[ma_ios_notification_handler remove_handler] in 000041.o
  "_AVAudioSessionRouteChangeReasonKey", referenced from:
      -[ma_ios_notification_handler handle_route_change:] in 000041.o
  "_AudioComponentFindNext", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioComponentInstanceDispose", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioComponentInstanceNew", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioOutputUnitStart", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioOutputUnitStop", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitAddPropertyListener", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitGetProperty", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitGetPropertyInfo", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitInitialize", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitRender", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitSetProperty", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_OBJC_CLASS_$_AVAudioSession", referenced from:
      objc-class-ref in 000041.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I am on MacOS/amd64 box.

@gen2brain,

Now, I got 

```bash
/usr/local/go/pkg/tool/darwin_amd64/link: running /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang failed: exit status 1
Undefined symbols for architecture arm64:
  "_AudioComponentFindNext", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioComponentInstanceDispose", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioComponentInstanceNew", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioOutputUnitStart", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioOutputUnitStop", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitAddPropertyListener", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitGetProperty", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitGetPropertyInfo", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitInitialize", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitRender", referenced from:
      _ma_context_init__coreaudio in 000041.o
  "_AudioUnitSetProperty", referenced from:
      _ma_context_init__coreaudio in 000041.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I feel we nearly there, just something missing. I am on MacOS/amd64 and try to cross-compile to iPhone which is arm64, not sure which is missing here.

--- a/README.md
+++ b/README.md
@@ -21,5 +21,5 @@ Documentation on [GoDoc](https://godoc.org/github.com/gen2brain/malgo). Also che
 * Windows (WASAPI, DirectSound, WinMM)
 * Linux (PulseAudio, ALSA, JACK)
 * FreeBSD/NetBSD/OpenBSD (OSS/audio(4)/sndio)
-* macOS (CoreAudio)
+* macOS/iOS (CoreAudio)
 * Android (OpenSL|ES, AAudio)

--- a/miniaudio.go
+++ b/miniaudio.go
@@ -11,6 +11,8 @@ package malgo
 #cgo netbsd LDFLAGS: -ldl -lpthread -lm
 #cgo freebsd LDFLAGS: -ldl -lpthread -lm
 #cgo android LDFLAGS: -lm
+#cgo ios CFLAGS: -x objective-c
+#cgo ios LDFLAGS: -framework CoreFoundation -framework AVFAudio -framework CoreAudio -framework AudioToolbox
 
 #cgo !noasm,!arm,!arm64 CFLAGS: -msse2
 #cgo !noasm,arm,arm64 CFLAGS: -mfpu=neon -mfloat-abi=hard

